conda/notices/main.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Module that provides the unofficial API for this feature. In order to maintain
+a good level of encapsulation, other parts of the application should only import
+from this module.
+"""
+
+import sys
+from functools import wraps
+from typing import Sequence, Tuple, Optional, Set
+from urllib import parse
+
+from conda.base.context import context, Context
+from conda.base.constants import NOTICES_FN
+from conda.models.channel import Channel
+
+from . import cache
+from . import views
+from . import http
+from .types import ChannelNotice, ChannelNoticeResponse
+
+
+def display_notices(
+    limit: Optional[int] = None,
+    always_show_viewed: bool = True,
+    silent: bool = False,
+) -> None:
+    """
+    Entry point for displaying notices. This is called by the "notices" decorator as well
+    as the sub-command "notices"
+    """
+    channel_name_urls = get_channel_name_and_urls(context.channel_objs)
+    channel_notice_responses = http.get_notice_responses(channel_name_urls, silent=silent)
+    channel_notices = flatten_notice_responses(channel_notice_responses)
+    num_total_notices = len(channel_notices)
+
+    cache_file = cache.get_notices_cache_file()
+    viewed_notices = None
+    num_viewed_notices = 0
+    if not always_show_viewed:
+        viewed_notices = cache.get_viewed_channel_notice_ids(cache_file, channel_notices)
+        num_viewed_notices = len(viewed_notices)
+
+    channel_notices = filter_notices(channel_notices, limit=limit, exclude=viewed_notices)
+    if len(channel_notices) == 0:
+        return
+
+    views.print_notices(channel_notices)
+
+    # Updates cache database, marking displayed notices as "viewed"
+    cache.mark_channel_notices_as_viewed(cache_file, channel_notices)
+
+    views.print_more_notices_message(num_total_notices, len(channel_notices), num_viewed_notices)
+
+
+NOTICES_DECORATOR_CONFIG_ERROR = (
+    "Unable to parse decorated function arguments for conda.notices."
+    ' Please make sure the function being decorated accepts both "args"'
+    ' and "parser" positional parameters'
+)
+
+
+def notices(func):
+    """
+    Wrapper for "execute" entry points for subcommands.
+
+    This decorator assumes two positional arguments:
+        - args
+        - parser
+
+    If this decorator is not configured correctly, we do our best to provide a friendly
+    error message.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            cmd_args, _ = args
+        except ValueError:
+            print(NOTICES_DECORATOR_CONFIG_ERROR, file=sys.stderr)
```

Comment:
I think this was perhaps a bit too defense coding. It was meant to be an error message helping developers if they were somehow to use the decorator incorrectly. I'm going to think about how this may be handled better including using and error logger.

Reply:
Sounds good to me, I was mostly curious, don't see a big deal tbh.